sockeye/generate_decoder_states.py

```diff
@@ -0,0 +1,239 @@
+        self.num_dim = num_dim  # dimension of a single entry
+        self.dtype = dtype
+        self.block_size = -1
+        self.mmap = None
+        self.tail_idx = 0  # where the next entry should be inserted
+        self.size = 0  # size of storage already assigned
+
+    @abstractmethod
+    def open(self, initial_size: int, block_size: int) -> None:
+        pass
+
+    @abstractmethod
+    def add(self, array: np.ndarray) -> None:
+        pass
+
+
+class NumpyMemmapStorage(BlockMMapStorage):
+    
+    def open(self, initial_size: int, block_size: int) -> None:
+        self.mmap = np.memmap(self.file_name, dtype=self.dtype, mode='w+', shape=(initial_size, self.num_dim))
+        self.size = initial_size
+        self.block_size = block_size
+
+    def add(self, array: np.ndarray) -> None:
+        """
+        It turns out that numpy memmap actually cannot be re-sized.
+        So we have to pre-estimate how many entries we need and put it down as initial_size.
+        If we end up adding more entries to the memmap than initially claimed, we'll have to bail out.
+        """
+        assert self.mmap is not None
+        num_entries, num_dim = array.shape
+        assert num_dim == self.num_dim
+
+        if self.tail_idx + num_entries > self.size:
+            # bail out
+            logger.warning(
+                "Trying to write {0} entries into a numpy memmap that has size {1} and already has {2} entries. Nothing is written."
+                .format(num_entries, self.size, self.tail_idx)
+            )
+        else:
+            start = self.tail_idx
+            end = self.tail_idx + num_entries
+            self.mmap[start:end] = array
+
+            self.tail_idx += num_entries
+
+
+class DecoderStateGenerator:
+
+    def __init__(self,
+                 model: SockeyeModel,
+                 source_vocabs: List[Vocab],
+                 target_vocabs: List[Vocab],
+                 output_prefix: str,
+                 max_seq_len_source: int,
+                 max_seq_len_target: int,
+                 state_data_type: str,
+                 word_data_type: str,
+                 device: pt.device) -> None:
+        self.model = model
+        self.source_vocabs = source_vocabs
+        self.target_vocabs = target_vocabs
+        self.device = device
+        self.traced_model = None
+        self.max_seq_len_source = max_seq_len_source
+        self.max_seq_len_target = max_seq_len_target
+
+        self.output_prefix = output_prefix
+        self.state_store_file = None
+        self.words_store_file = None
+
+        # info for KNNConfig
+        self.num_states = 0
+        self.dimension = None
+        self.state_data_type = utils.get_numpy_dtype(state_data_type)
+        self.word_data_type = utils.get_numpy_dtype(word_data_type)
+
+    @staticmethod
+    def probe_token_count(target: str, max_seq_len: int) -> int:
+        token_count = 0
+        with codecs.open(target, 'r', 'utf-8') as f:
```

Comment:
Why do we need `codecs.open` here? The default encoding for standard `open` should be utf-8  already in recent versions of Python.

Reply:
I've had encoding issues with `open` in Python 3.6 but probably with 3.7 and 3.8 it's fixed? And if that's the case, can we assume people will only run Sockeye in Python 3.7+?

Reply:
Yes, we require Python3.7+ (https://awslabs.github.io/sockeye/setup.html)